domain/application/state/application.go

```diff
@@ -3014,6 +3016,282 @@ WHERE application_uuid = $applicationID.uuid;
 	return hash.SHA256, nil
 }
 
+// GetApplicationConstraints returns the application constraints for the
+// specified application ID.
+// Empty constraints are returned if no constraints exist for the given
+// application ID.
+func (st *State) GetApplicationConstraints(ctx context.Context, appID coreapplication.ID) (constraints.Value, error) {
+	db, err := st.DB()
+	if err != nil {
+		return constraints.Value{}, internalerrors.Capture(err)
+	}
+
+	ident := applicationID{ID: appID}
+
+	query := `
+SELECT &applicationConstraint.*
+FROM application_constraint
+JOIN "constraint" ON "constraint".uuid = application_constraint.constraint_uuid
+LEFT JOIN container_type ON container_type.id = "constraint".container_type_id
+LEFT JOIN constraint_tag ON constraint_tag.constraint_uuid = "constraint".uuid	
+LEFT JOIN constraint_space ON constraint_space.constraint_uuid = "constraint".uuid	
+LEFT JOIN constraint_zone ON constraint_zone.constraint_uuid = "constraint".uuid	
+WHERE application_uuid = $applicationID.uuid;
+`
+
+	stmt, err := st.Prepare(query, applicationConstraint{}, ident)
+	if err != nil {
+		return constraints.Value{}, internalerrors.Errorf("preparing query for application constraints: %w", err)
+	}
+
+	var result applicationConstraints
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, ident).GetAll(&result)
```

Comment:
Our `ident` is the app ID, so should be a unique specifier. So just a `Get` should be sufficient.

Reply:
Then you should handle the `sqlair.ErrNoRows` case by returning early

Reply:
No. The constraints have embedded slices (spaces, tags and zones), so the results is a list of rows with flat values.

Reply:
Hmm good point. In that case, I'm a little unsure if a join is the best way to extract this data, but I don't know really. The joined tables don't really represent anything. 

An alternative would be separate queries? Query the constraint, and then separately query the tag, space and zone?

Although I might be starting bike-shed here

What do you think? @SimonRichardson

Reply:
I can tell you 100% confidently that joining tables is the way to go. This is classic relational algebra. Having embedded results coming from different tables is OK and we just have to flatten the results.

Reply:
There is a point here... if the constraints are compounded, what happens to the old constraints.

Assume:

 1. insert arch and cpu-core count
 2. insert virt-type
 3. insert cpu-core count

What happens to the cpu-core count? We should remove the old one, right? Or is it the last write wins? Yet we don't want this because it can lead to unbounded growth, and we know that dqlite can't handle that.

If we have a strategy for this, then I'm fine. Otherwise, I tend to agree with @jack-w-shaw and that we should have separate tables for slices.

Reply:
> What happens to the cpu-core count? 

What I've implemented is what we already have on 3.6: every time we set constraints it's a full rewrite. So in your example we'll only have cpu-core count, and the value provided in step 3. 

> I tend to agree with @jack-w-shaw and that we should have separate tables for slices.

What do you mean? We already have separate tables. See https://github.com/juju/juju/blob/main/domain/schema/model/sql/0014-constraint.sql#L40-L69